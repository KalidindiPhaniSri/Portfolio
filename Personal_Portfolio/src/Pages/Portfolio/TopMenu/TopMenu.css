.top-menu {
  height: 4rem;
}

.top-menu .appbar-nav {
  background-color: inherit !important;
  padding: 0 4rem;
}
