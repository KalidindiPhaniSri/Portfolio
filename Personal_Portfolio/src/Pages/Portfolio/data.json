{
  "topMenu": {
    "text": "Phani",
    "icon": "",
    "sections": ["Home", "About", "Skills", "Experience", "Projects", "Contact"]
  },
  "home": {
    "bio": {
      "intro": "Hello, I am",
      "name": "Phani Sri Kalidindi",
      "role": "Full Stack Developer",
      "text": "Full Stack Developer with 2 years of experience building scalable web apps using React.js, TypeScript, .NET Core, and SQL. Passionate about clean architecture and solving real-world problems."
    }
  },
  "about": {
    "title": ["About", "Me"],
    "text": "I'm a Full Stack Developer with 2 years of experience building scalable, data-driven web applications. Iâ€™ve developed enterprise tools for Walmart, including Supply Chain and Deduction Management systems. I'm skilled in React.js, TypeScript, JavaScript (ES6+), Redux, Material UI, and Kendo UI. I integrate RESTful APIs, optimize performance with lazy loading and code splitting, and implement secure Auth0 authentication. On the backend, I'm familiar with C#, .NET Core, and SQL, and I focus on clean, maintainable code with unit testing. I enjoy solving complex problems and delivering real business value in Agile teams."
  }
}
