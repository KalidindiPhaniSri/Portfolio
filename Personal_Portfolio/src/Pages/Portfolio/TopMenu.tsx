import * as React from "react";
import AppBar from "@mui/material/AppBar";
import Box from "@mui/material/Box";
import Divider from "@mui/material/Divider";
import Drawer from "@mui/material/Drawer";
import IconButton from "@mui/material/IconButton";
import List from "@mui/material/List";
import ListItem from "@mui/material/ListItem";
import ListItemButton from "@mui/material/ListItemButton";
import ListItemText from "@mui/material/ListItemText";
import MenuIcon from "@mui/icons-material/Menu";
import Toolbar from "@mui/material/Toolbar";
import Typography from "@mui/material/Typography";
import Button from "@mui/material/Button";
import data from "./data.json";

interface TopMenuProps {
  activeSection: string;
  scrolled: boolean;
}

const drawerWidth = 240;

const getNavBtnClass = (
  item: string,
  activeSection: string,
  scrolled: boolean
) => {
  if (activeSection.toLowerCase() === item.toLowerCase())
    return "highlight-navbtn";
  if (scrolled) return "scrolled-navbtn";
  return "navbtn";
};

const TopMenu: React.FC<TopMenuProps> = ({ activeSection, scrolled }) => {
  const [mobileOpen, setMobileOpen] = React.useState(false);
  const { text, sections } = data.topMenu;
  const handleDrawerToggle = () => setMobileOpen((prev) => !prev);

  React.useEffect(() => {
    const handleAnchorClick = (e: Event) => {
      const target = e.target as HTMLAnchorElement;
      if (
        target.tagName === "A" &&
        target.getAttribute("href")?.startsWith("#")
      ) {
        const id = target.getAttribute("href")!.slice(1);
        const el = document.getElementById(id);
        if (el) {
          e.preventDefault();
          // Dynamically get the nav height
          const nav = document.querySelector(".top-menu");
          const navHeight = nav ? (nav as HTMLElement).offsetHeight : 0;
          const rect = el.getBoundingClientRect();
          const scrollTop =
            window.pageYOffset || document.documentElement.scrollTop;
          const top = rect.top + scrollTop - navHeight;
          window.scrollTo({ top, behavior: "smooth" });
        }
      }
    };
    document.addEventListener("click", handleAnchorClick);
    return () => document.removeEventListener("click", handleAnchorClick);
  }, []);

  const renderDrawerList = () => (
    <List>
      {sections.map((item) => (
        <ListItem key={item} disablePadding>
          <ListItemButton
            sx={{ textAlign: "center" }}
            component="a"
            href={`#${item.toLowerCase()}`}
          >
            <ListItemText
              primary={item}
              className={`${
                activeSection.toLowerCase() === item.toLowerCase()
                  ? "highlight-navbtn"
                  : ""
              }`}
            />
          </ListItemButton>
        </ListItem>
      ))}
    </List>
  );

  const drawer = (
    <Box
      onClick={handleDrawerToggle}
      sx={{ textAlign: "center" }}
      data-testid="mobile-nav"
    >
      <Typography variant="h6" sx={{ my: 2 }}>
        MUI
      </Typography>
      <Divider />
      {renderDrawerList()}
    </Box>
  );

  return (
    <Box className={`top-menu${scrolled ? " scrolled" : ""}`}>
      <AppBar
        component="nav"
        className="appbar-nav"
        sx={{ paddingInline: { xs: "2rem", md: "4rem" } }}
      >
        <Toolbar>
          <IconButton
            color="inherit"
            aria-label="open drawer"
            edge="start"
            onClick={handleDrawerToggle}
            sx={{
              mr: 2,
              display: { xs: "block", md: "none" },
              color: "var(--theme-yellow)",
            }}
          >
            <MenuIcon fontSize="large" />
          </IconButton>
          <Typography
            variant="h6"
            component="div"
            sx={{
              flexGrow: 1,
              display: { xs: "none", md: "block" },
              color: "var(--theme-yellow)",
            }}
          >
            {text}
          </Typography>
          <Box
            sx={{ display: { xs: "none", md: "block" } }}
            data-testid="desktop-nav"
          >
            {sections.map((item) => (
              <Button
                key={item}
                href={`#${item.toLowerCase()}`}
                className={getNavBtnClass(item, activeSection, scrolled)}
              >
                {item}
              </Button>
            ))}
          </Box>
        </Toolbar>
      </AppBar>
      <nav>
        <Drawer
          variant="temporary"
          open={mobileOpen}
          onClose={handleDrawerToggle}
          ModalProps={{
            keepMounted: true,
          }}
          sx={{
            display: { xs: "block", md: "none" },
            "& .MuiDrawer-paper": {
              boxSizing: "border-box",
              width: drawerWidth,
            },
          }}
        >
          {drawer}
        </Drawer>
      </nav>
      <Box component="main">
        <Toolbar />
      </Box>
    </Box>
  );
};

export default TopMenu;
