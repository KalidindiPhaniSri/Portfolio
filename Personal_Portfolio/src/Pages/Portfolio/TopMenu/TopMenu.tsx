import * as React from "react";
import AppBar from "@mui/material/AppBar";
import Box from "@mui/material/Box";
import Divider from "@mui/material/Divider";
import Drawer from "@mui/material/Drawer";
import IconButton from "@mui/material/IconButton";
import List from "@mui/material/List";
import ListItem from "@mui/material/ListItem";
import ListItemButton from "@mui/material/ListItemButton";
import ListItemText from "@mui/material/ListItemText";
import MenuIcon from "@mui/icons-material/Menu";
import Toolbar from "@mui/material/Toolbar";
import Typography from "@mui/material/Typography";
import Button from "@mui/material/Button";
import data from "../data.json";
import "./TopMenu.css";

interface TopMenuProps {
  activeSection: string;
}

const drawerWidth = 240;

const TopMenu: React.FC<TopMenuProps> = ({ activeSection }) => {
  const [mobileOpen, setMobileOpen] = React.useState(false);
  const { text, sections } = data.topMenu;

  const handleDrawerToggle = () => {
    setMobileOpen((prevState) => !prevState);
  };

  const drawer = (
    <Box
      onClick={handleDrawerToggle}
      sx={{ textAlign: "center" }}
      data-testid="mobile-nav"
    >
      <Typography variant="h6" sx={{ my: 2 }}>
        MUI
      </Typography>
      <Divider />
      <List>
        {data.topMenu.sections.map((item) => (
          <ListItem key={item} disablePadding>
            <ListItemButton
              sx={{ textAlign: "center" }}
              component="a"
              href={"#" + item.toLowerCase()}
            >
              <ListItemText
                primary={item}
                sx={
                  activeSection.toLowerCase() === item.toLowerCase()
                    ? { color: "#f3bc17" }
                    : {}
                }
              />
            </ListItemButton>
          </ListItem>
        ))}
      </List>
    </Box>
  );

  return (
    <Box sx={{ display: "flex" }} className="top-menu">
      <AppBar component="nav" className="appbar-nav">
        <Toolbar>
          <IconButton
            color="inherit"
            aria-label="open drawer"
            edge="start"
            onClick={handleDrawerToggle}
            sx={{ mr: 2, display: { md: "none" } }}
          >
            <MenuIcon />
          </IconButton>
          <Typography
            variant="h6"
            component="div"
            sx={{
              flexGrow: 1,
              display: { xs: "none", md: "block" },
              color: "#f3bc17",
            }}
          >
            {text}
          </Typography>
          <Box
            sx={{ display: { xs: "none", md: "block" } }}
            data-testid="desktop-nav"
          >
            {sections.map((item) => (
              <Button
                key={item}
                sx={
                  activeSection.toLowerCase() === item.toLowerCase()
                    ? { color: "#f3bc17" }
                    : { color: "#fff" }
                }
                href={"#" + item.toLowerCase()}
              >
                {item}
              </Button>
            ))}
          </Box>
        </Toolbar>
      </AppBar>
      <nav>
        <Drawer
          variant="temporary"
          open={mobileOpen}
          onClose={handleDrawerToggle}
          ModalProps={{
            keepMounted: true, // Better open performance on mobile.
          }}
          sx={{
            display: { xs: "block", md: "none" },
            "& .MuiDrawer-paper": {
              boxSizing: "border-box",
              width: drawerWidth,
            },
          }}
        >
          {drawer}
        </Drawer>
      </nav>
      <Box component="main">
        <Toolbar />
      </Box>
    </Box>
  );
};
export default TopMenu;
