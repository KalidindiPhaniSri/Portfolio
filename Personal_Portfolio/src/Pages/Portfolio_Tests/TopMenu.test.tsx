import { fireEvent, render, screen, within } from "@testing-library/react";
import { describe, expect } from "vitest";
import data from "./testdata.json";
import TopMenu from "../Portfolio/TopMenu";

describe("Top Menu", () => {
  it("renders logo text", () => {
    render(<TopMenu activeSection="home" scrolled={false} />);
    expect(screen.getByText(data.topMenu.text)).toBeInTheDocument();
  });

  it("render all the nav items with href in desktop mode", () => {
    render(<TopMenu activeSection="home" scrolled={false} />);
    const desktopNav = screen.getByTestId("desktop-nav");
    data.topMenu.sections.forEach((item) => {
      const link = within(desktopNav).getByText(item);
      expect(link).toBeInTheDocument();
      expect(link).toHaveAttribute("href", `#${item.toLowerCase()}`);
    });
  });

  it("opening and closing the drawer ", () => {
    global.innerWidth = 500;
    global.dispatchEvent(new Event("resize"));

    render(<TopMenu activeSection="home" scrolled={false} />);
    const menuButton = screen.getByLabelText("open drawer");
    expect(menuButton).toBeInTheDocument();

    fireEvent.click(menuButton);
    const mobileNav = screen.getByTestId("mobile-nav");
    data.topMenu.sections.forEach((item) => {
      const textNode = within(mobileNav).getByText(item);
      const anchorNode = textNode.closest("a");
      expect(textNode).toBeInTheDocument();
      expect(anchorNode).toHaveAttribute("href", `#${item.toLowerCase()}`);
    });
    fireEvent.click(menuButton);
    expect(mobileNav).not.toBeVisible();
  });
});
// Passing tests
describe("Top Menu - Passing Tests", () => {
  it("applies highlight-navbtn class to active section in desktop", () => {
    render(
      <TopMenu activeSection={data.topMenu.sections[1]} scrolled={false} />
    );
    const desktopNav = screen.getByTestId("desktop-nav");
    const activeBtn = within(desktopNav).getByText(data.topMenu.sections[1]);
    expect(activeBtn.className).toContain("highlight-navbtn");
  });

  it("applies scrolled-navbtn class to non-active sections when scrolled", () => {
    render(
      <TopMenu activeSection={data.topMenu.sections[0]} scrolled={true} />
    );
    const desktopNav = screen.getByTestId("desktop-nav");
    data.topMenu.sections.slice(1).forEach((item) => {
      const btn = within(desktopNav).getByText(item);
      expect(btn.className).toContain("scrolled-navbtn");
    });
  });

  it("check menu icon", () => {
    global.innerWidth = 500;
    global.dispatchEvent(new Event("resize"));
    render(<TopMenu activeSection="home" scrolled={false} />);
    expect(screen.getByLabelText("open drawer")).toBeInTheDocument();
  });
});
