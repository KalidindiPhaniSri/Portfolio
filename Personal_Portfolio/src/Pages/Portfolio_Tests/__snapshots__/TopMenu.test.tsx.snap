// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Top Menu - Additional Cases > matches snapshot for default state 1`] = `
<DocumentFragment>
  <div
    class="top-menu MuiBox-root css-0"
  >
    <nav
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed appbar-nav mui-fixed css-wy9qcc-MuiPaper-root-MuiAppBar-root"
      style="--Paper-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1ygil4i-MuiToolbar-root"
      >
        <button
          aria-label="open drawer"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeMedium css-1uexjlw-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-uks7c6-MuiSvgIcon-root"
            data-testid="MenuIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"
            />
          </svg>
        </button>
        <div
          class="MuiTypography-root MuiTypography-h6 css-pj280i-MuiTypography-root"
        >
          Phani
        </div>
        <div
          class="MuiBox-root css-13tqxrv"
          data-testid="desktop-nav"
        >
          <a
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary highlight-navbtn css-1uent87-MuiButtonBase-root-MuiButton-root"
            href="#home"
            tabindex="0"
          >
            Home
          </a>
          <a
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary navbtn css-1uent87-MuiButtonBase-root-MuiButton-root"
            href="#about"
            tabindex="0"
          >
            About
          </a>
          <a
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary navbtn css-1uent87-MuiButtonBase-root-MuiButton-root"
            href="#skills"
            tabindex="0"
          >
            Skills
          </a>
          <a
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary navbtn css-1uent87-MuiButtonBase-root-MuiButton-root"
            href="#experience"
            tabindex="0"
          >
            Experience
          </a>
          <a
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary navbtn css-1uent87-MuiButtonBase-root-MuiButton-root"
            href="#projects"
            tabindex="0"
          >
            Projects
          </a>
          <a
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary navbtn css-1uent87-MuiButtonBase-root-MuiButton-root"
            href="#contact"
            tabindex="0"
          >
            Contact
          </a>
        </div>
      </div>
    </nav>
    <nav />
    <main
      class="MuiBox-root css-0"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1ygil4i-MuiToolbar-root"
      />
    </main>
  </div>
</DocumentFragment>
`;
